# Dealing with Date and Time Pandas

In [1]:
import pandas as pd
from pandas.api.types import CategoricalDtype
from datetime import datetime

In [2]:
df = pd.read_csv("data/AAPL.csv")
df

Date        Open        High         Low       Close   Adj Close  \
0     1980-12-12    0.513393    0.515625    0.513393    0.513393    0.407747   
1     1980-12-15    0.488839    0.488839    0.486607    0.486607    0.386473   
2     1980-12-16    0.453125    0.453125    0.450893    0.450893    0.358108   
3     1980-12-17    0.462054    0.464286    0.462054    0.462054    0.366972   
4     1980-12-18    0.475446    0.477679    0.475446    0.475446    0.377609   
...          ...         ...         ...         ...         ...         ...   
9818  2019-11-19  267.899994  268.000000  265.390015  266.290009  266.290009   
9819  2019-11-20  265.540009  266.079987  260.399994  263.190002  263.190002   
9820  2019-11-21  263.690002  264.010010  261.179993  262.010010  262.010010   
9821  2019-11-22  262.589996  263.179993  260.839996  261.779999  261.779999   
9822  2019-11-25  262.709991  266.440002  262.519989  266.369995  266.369995   

           Volume  
0     117258400.0  
1      43971200.0  
2      26432000.0  
3      21610400.0  
4      18362400.0  
...           ...  
9818   19041800.0  
9819   26558600.0  
9820   30348800.0  
9821   16331300.0  
9822   21005100.0  

[9823 rows x 7 columns]

You can examine the data type of each column in the DataFrame:

In [3]:
df.dtypes

Date          object
Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume       float64
dtype: object

---

You can convert the Date column to the datetime64 data type using the datetime.strptime() function

In [4]:
df["Date"] = df["Date"].apply(lambda x: datetime.strptime(x, "%Y-%m-%d"))
df

Date        Open        High         Low       Close   Adj Close  \
0    1980-12-12    0.513393    0.515625    0.513393    0.513393    0.407747   
1    1980-12-15    0.488839    0.488839    0.486607    0.486607    0.386473   
2    1980-12-16    0.453125    0.453125    0.450893    0.450893    0.358108   
3    1980-12-17    0.462054    0.464286    0.462054    0.462054    0.366972   
4    1980-12-18    0.475446    0.477679    0.475446    0.475446    0.377609   
...         ...         ...         ...         ...         ...         ...   
9818 2019-11-19  267.899994  268.000000  265.390015  266.290009  266.290009   
9819 2019-11-20  265.540009  266.079987  260.399994  263.190002  263.190002   
9820 2019-11-21  263.690002  264.010010  261.179993  262.010010  262.010010   
9821 2019-11-22  262.589996  263.179993  260.839996  261.779999  261.779999   
9822 2019-11-25  262.709991  266.440002  262.519989  266.369995  266.369995   

           Volume  
0     117258400.0  
1      43971200.0  
2      26432000.0  
3      21610400.0  
4      18362400.0  
...           ...  
9818   19041800.0  
9819   26558600.0  
9820   30348800.0  
9821   16331300.0  
9822   21005100.0  

[9823 rows x 7 columns]

In [5]:
df.dtypes

Date         datetime64[ns]
Open                float64
High                float64
Low                 float64
Close               float64
Adj Close           float64
Volume              float64
dtype: object

---

You can obtain the list of format codes (such as %Y, %m, and %d) from https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes.

The [ns] in datetime64[ns] specifies the precision of the DateTime object in nano second.

## Finding Rows Based on Specific Dates

Converting the Date column to datetime64 data type allows you to perform date-related operations easily, such as finding all the rows for December 2018

In [6]:
df[(df["Date"].dt.month == 12) & (df["Date"].dt.year == 2018)]

Date        Open        High         Low       Close   Adj Close  \
9576 2018-12-03  184.460007  184.940002  181.210007  184.820007  182.084183   
9577 2018-12-04  180.949997  182.389999  176.270004  176.690002  174.074539   
9578 2018-12-06  171.759995  174.779999  170.419998  174.720001  172.133682   
9579 2018-12-07  173.490005  174.490005  168.300003  168.490005  165.995926   
9580 2018-12-10  165.000000  170.089996  163.330002  169.600006  167.089478   
9581 2018-12-11  171.660004  171.789993  167.000000  168.630005  166.133850   
9582 2018-12-12  170.399994  171.919998  169.020004  169.100006  166.596878   
9583 2018-12-13  170.490005  172.570007  169.550003  170.949997  168.419479   
9584 2018-12-14  169.000000  169.080002  165.279999  165.479996  163.030457   
9585 2018-12-17  165.449997  168.350006  162.729996  163.940002  161.513260   
9586 2018-12-18  165.380005  167.529999  164.389999  166.070007  163.611740   
9587 2018-12-19  166.000000  167.449997  159.089996  160.889999  158.508408   
9588 2018-12-20  160.399994  162.110001  155.300003  156.830002  154.508499   
9589 2018-12-21  156.860001  158.160004  149.630005  150.729996  148.498795   
9590 2018-12-24  148.149994  151.550003  146.589996  146.830002  144.656540   
9591 2018-12-26  148.300003  157.229996  146.720001  157.169998  154.843475   
9592 2018-12-27  155.839996  156.770004  150.070007  156.149994  153.838562   
9593 2018-12-28  157.500000  158.520004  154.550003  156.229996  153.917389   
9594 2018-12-31  158.529999  159.360001  156.479996  157.740005  155.405045   

          Volume  
9576  40802500.0  
9577  41344300.0  
9578  43098400.0  
9579  42281600.0  
9580  62026000.0  
9581  47281700.0  
9582  35627700.0  
9583  31898600.0  
9584  40703700.0  
9585  44287900.0  
9586  33841500.0  
9587  49047300.0  
9588  64773000.0  
9589  95744600.0  
9590  37169200.0  
9591  58582500.0  
9592  53117100.0  
9593  42291400.0  
9594  35003500.0

## Finding Rows Based on Day of Week

You can also use the dayofweek attribute to find a particular day in the week (such as Monday, Tuesday, and so on). For example, the following statement finds all the rows whose dates fall on a Monday:

In [7]:
# 0 is Monday, 1 is Tue, etc
df[df["Date"].dt.dayofweek == 0].sample(5)

Date        Open        High         Low       Close   Adj Close  \
2622 1991-04-29    2.089286    2.151786    2.080357    2.080357    1.715720   
2399 1990-06-11    1.348214    1.392857    1.348214    1.392857    1.139037   
8657 2015-04-13  128.369995  128.570007  126.610001  126.849998  117.011948   
2773 1991-12-02    1.812500    1.857143    1.785714    1.848214    1.535315   
1513 1986-12-08    0.779018    0.783482    0.756696    0.758929    0.602757   

          Volume  
2622  51676800.0  
2399  39474400.0  
8657  36365100.0  
2773  29724800.0  
1513  86800000.0

## Sorting by Day of Week

Another common scenario you will encounter when dealing with dates is sorting the data based on the day of week (such as from Monday to Sunday, or from Sunday to Saturday).

Here’s a technique that you can use. First, extract the day of week from the Date column (a datetime64[ns] object) using the strftime() function:

In [8]:
df["Day"] = df["Date"].apply(lambda x: x.strftime("%a"))
df

Date        Open        High         Low       Close   Adj Close  \
0    1980-12-12    0.513393    0.515625    0.513393    0.513393    0.407747   
1    1980-12-15    0.488839    0.488839    0.486607    0.486607    0.386473   
2    1980-12-16    0.453125    0.453125    0.450893    0.450893    0.358108   
3    1980-12-17    0.462054    0.464286    0.462054    0.462054    0.366972   
4    1980-12-18    0.475446    0.477679    0.475446    0.475446    0.377609   
...         ...         ...         ...         ...         ...         ...   
9818 2019-11-19  267.899994  268.000000  265.390015  266.290009  266.290009   
9819 2019-11-20  265.540009  266.079987  260.399994  263.190002  263.190002   
9820 2019-11-21  263.690002  264.010010  261.179993  262.010010  262.010010   
9821 2019-11-22  262.589996  263.179993  260.839996  261.779999  261.779999   
9822 2019-11-25  262.709991  266.440002  262.519989  266.369995  266.369995   

           Volume  Day  
0     117258400.0  Fri  
1      43971200.0  Mon  
2      26432000.0  Tue  
3      21610400.0  Wed  
4      18362400.0  Thu  
...           ...  ...  
9818   19041800.0  Tue  
9819   26558600.0  Wed  
9820   30348800.0  Thu  
9821   16331300.0  Fri  
9822   21005100.0  Mon  

[9823 rows x 8 columns]

You then define the order of the dates that you want to order:

In [9]:
# ---define your category order---
cats = ["Mon", "Tue", "Wed", "Thu", "Fri"]  # omit Sat and Sun

In the above I omitted the Saturday and Sunday since there is no stock data on these two days (weekends).

Then, create a new categorical data type using the CategoricalDtype class by passing the order of the days that you have defined earlier:

In [10]:
cat_type = CategoricalDtype(categories=cats, ordered=True)

Finally, convert the Day column into the new categorical type that you have just created:

In [11]:
# ---cast the Day column as categorical---
df["Day"] = df["Day"].astype(cat_type)
df.dtypes

Date         datetime64[ns]
Open                float64
High                float64
Low                 float64
Close               float64
Adj Close           float64
Volume              float64
Day                category
dtype: object

You can now see that Day is of category data type:

If you want to print out all the stock data sorted by the day of week, you can now use a groupby() function:

In [12]:
for _, gp in df.groupby("Day"):
    display(gp.sample(3))

Date       Open       High        Low      Close  Adj Close  \
1600 1987-04-13   1.250000   1.254464   1.205357   1.205357   0.957319   
8484 2014-08-04  96.370003  96.580002  95.169998  95.589996  87.017715   
2690 1991-08-05   1.776786   1.776786   1.723214   1.732143   1.432200   

          Volume  Day  
1600  35554400.0  Mon  
8484  39958000.0  Mon  
2690  25191600.0  Mon

Date        Open        High         Low       Close   Adj Close  \
8605 2015-01-27  112.419998  112.480003  109.029999  109.139999  100.279716   
1155 1985-07-09    0.314732    0.316964    0.314732    0.314732    0.249967   
9363 2018-01-30  165.529999  167.369995  164.699997  166.970001  162.065277   

          Volume  Day  
8605  95568700.0  Tue  
1155  36976800.0  Tue  
9363  46048200.0  Tue

Date        Open        High         Low       Close   Adj Close  \
6166 2005-05-18    5.064286    5.365714    4.998571    5.120000    4.454917   
6455 2006-07-12    7.881429    7.891428    7.560000    7.565714    6.582935   
9672 2019-04-24  207.360001  208.479996  207.050003  207.160004  204.968811   

           Volume  Day  
6166  159180700.0  Wed  
6455  231832300.0  Wed  
9672   17540600.0  Wed

Date      Open      High       Low     Close  Adj Close  \
5887 2004-04-08  1.991429  2.000000  1.942857  1.966429   1.710991   
469  1982-10-21  0.453125  0.477679  0.446429  0.464286   0.368745   
6007 2004-09-30  2.785714  2.805000  2.746428  2.767857   2.408315   

           Volume  Day  
5887   60229400.0  Thu  
469    56879200.0  Thu  
6007  106253000.0  Thu

Date       Open       High        Low      Close  Adj Close  \
2689 1991-08-02   1.776786   1.794643   1.750000   1.785714   1.476495   
5853 2004-02-20   1.607143   1.607857   1.586429   1.600000   1.392162   
7988 2012-08-10  88.387146  88.822861  88.385712  88.814285  77.609184   

          Volume  Day  
2689  68252800.0  Fri  
5853  69400800.0  Fri  
7988  48734700.0  Fri

Observe that the groups are sorted by Day, with Monday first, followed by Tue, etc:

If you want Tues to be displayed first, followed by Wed and so on, change the order in the cats variable:

In [13]:
cats = ["Tue", "Wed", "Thu", "Fri", "Mon"]  # omit Sat and Sun
cat_type = CategoricalDtype(categories=cats, ordered=True)

# ---cast the Day column as categorical---
df["Day"] = df["Day"].astype(cat_type)

for _, gp in df.groupby("Day"):
    display(gp.sample(3))

Date       Open       High        Low      Close  Adj Close  \
3341 1994-03-01   1.312500   1.312500   1.276786   1.294643   1.102197   
5740 2003-09-09   1.609286   1.619286   1.580000   1.597857   1.390297   
7054 2008-11-25  13.518572  13.530000  12.594286  12.971429  11.286451   

           Volume  Day  
3341   52967600.0  Tue  
5740   45092600.0  Tue  
7054  308823200.0  Tue

Date       Open       High        Low      Close  Adj Close  \
4968 2000-08-09   3.437500   3.459821   3.375000   3.392857   2.952128   
6604 2007-02-14  12.090000  12.234285  12.081429  12.185715  10.602801   
904  1984-07-11   0.479911   0.486607   0.466518   0.473214   0.375836   

           Volume  Day  
4968   94910200.0  Wed  
6604  126995400.0  Wed  
904    30273600.0  Wed

Date      Open      High       Low     Close  Adj Close  \
3963 1996-08-15  0.808036  0.812500  0.794643  0.794643   0.691420   
3711 1995-08-17  1.593750  1.625000  1.575893  1.593750   1.382411   
1583 1987-03-19  1.174107  1.223214  1.169643  1.220982   0.969729   

          Volume  Day  
3963  26905200.0  Thu  
3711  61723200.0  Thu  
1583  51682400.0  Thu

Date        Open        High         Low       Close   Adj Close  \
3402 1994-05-27    1.080357    1.098214    1.053571    1.069196    0.913862   
9385 2018-03-02  172.800003  176.300003  172.449997  176.210007  171.731186   
9039 2016-10-14  117.879997  118.169998  117.129997  117.629997  111.793686   

          Volume  Day  
3402  27171200.0  Fri  
9385  38454000.0  Fri  
9039  35652200.0  Fri

Date        Open        High         Low       Close   Adj Close  \
9106 2017-01-23  120.000000  120.809998  119.769997  120.080002  114.708061   
9045 2016-10-24  117.099998  117.739998  117.000000  117.650002  111.812714   
1779 1987-12-28    1.508929    1.517857    1.410714    1.437500    1.147401   

          Volume  Day  
9106  22050200.0  Mon  
9045  23538700.0  Mon  
1779  57400000.0  Mon

## Converting DateTime Columns During Loading Time

In the earlier section you converted the Date column to the datetime64 data type after the entire CSV file has been loaded into the DataFrame. Another way of doing the same thing would be to perform the conversion while you are loading the CSV using the parse_dates parameter:

In [14]:
df = pd.read_csv("data/AAPL.csv", parse_dates=["Date"])

The parse_dates parameter specifies the column(s) to parse as a datetime64 object.

> Note that if the parse_dates parameter is set to True, Pandas will try to parse the index as an datetime64 object

## Parsing Multiple Columns as Date

Sometimes the dates in a dataset are stored separately in different columns, such as one column for year, one column for months, etc. While there are pros and cons of saving the data this way, it is sometimes easier if all the different columns can be combined into a single one. A good example of this representation is in the flights.csv file.

In [15]:
df = pd.read_csv("data/flights.csv", low_memory=False)

# display sample 5 rows and first 5 columns
df.sample(5).iloc[:, :5]

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE
5488065  2015     12   10            4      AA
2329500  2015      5   28            4      DL
818780   2015      2   23            1      EV
4915943  2015     11    3            2      WN
3131099  2015      7   15            3      WN

## Combining Columns during Loading Time

In the flights.csv file, the date of each flight is represented using three different columns — YEAR, MONTH, and DAY. For performing data analytics, it would be easier if you could combine the three column into a single date column using the parse_dates parameter:

In [16]:
df = pd.read_csv("data/flights.csv", parse_dates=[[0, 1, 2]], low_memory=False)
df.sample(5).iloc[:, :5]

YEAR_MONTH_DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER
1288742     2015-03-24            2      UA            633      N667UA
2547182     2015-06-10            3      UA           1170      N37422
2046109     2015-05-10            7      US           1885      N815AW
497775      2015-02-02            1      MQ           3068      N665MQ
3706960     2015-08-18            2      DL           1690      N951DL

In the above code snippet, I used the column index to indicate the columns to combine into a single column. I could also specify the name of the columns:

In [17]:
df = pd.read_csv(
    "data/flights.csv", parse_dates=[["YEAR", "MONTH", "DAY"]], low_memory=False
)

The result is that the first three columns are removed and replaced with a new column whose name is the concatenation of the three column names:

Notice that earlier in the article I set the parse_dates parameter with a list — parse_dates=[‘Date’]. Setting it to a list will cause the individual columns to be loaded as datetime objects. If you set it to a list of list, like parse_dates=[[‘YEAR’,’MONTH’,’DAY’]], then these columns will be combined into a single datetime object.

Assuming you have a CSV that looks like this:

    DATE,YEAR,MONTH,DAY
    2015-09-13,2015,9,13
    2015-09-14,2015,9,14

You can convert the DATE column to a datetime object and at the same time combine the YEAR, MONTH, and DAY columns into a single datetime object using the following values for the parse_dates parameter:

    df = pd.read_csv("test.csv", parse_dates=['DATE',['YEAR','MONTH','DAY']])

## Combining Columns After DataFrame is Loaded

You can also combine the columns of a DataFrame after it has been loaded using the to_datetime() function:

In [18]:
df = pd.read_csv("data/flights.csv", low_memory=False)
df["DATETIME"] = pd.to_datetime(df[["YEAR", "MONTH", "DAY"]])
df.sample(5).iloc[:, -3:]

LATE_AIRCRAFT_DELAY  WEATHER_DELAY   DATETIME
4469554                  NaN            NaN 2015-10-06
2229843                  NaN            NaN 2015-05-22
4910683                  NaN            NaN 2015-11-03
3083214                  NaN            NaN 2015-07-12
908820                  13.0            0.0 2015-03-01

The following output shows the last three columns of the DataFrame, with the last column being the result of combining the three columns — YEAR, MONTH, and DAY:

## Formatting Time

Besides storing the scheduled departure date in three separate columns, you would also notice that there is a column named SCHEDULED_DEPARTURE:

In [19]:
df[["SCHEDULED_DEPARTURE"]]

SCHEDULED_DEPARTURE
0                          5
1                         10
2                         20
3                         20
4                         25
...                      ...
5819074                 2359
5819075                 2359
5819076                 2359
5819077                 2359
5819078                 2359

[5819079 rows x 1 columns]

This column is an integer column that stores numbers such as 5,10, up to 2359. What each value represents is actually the time of departure in HHMM format. So 5 actually represents 00:05, while 2359 actually represents 23:59. If you were to perform analytics on this column, you definitely need to process this column further.

What I will do here is to combine the four columns into a datetime column:

- YEAR
- MONTH
- DAY
- SCHEDULED_DEPARTURE

Combining the first three is easy as we have seen in the previous section. The fourth column needs some processing:

- You need to format the departure time as a string and then extract the first 2 digits to represent the hour (HH)
- Then extract the last two digits representing the minutes (MM)

In [20]:
import datetime

# function to convert HHMM to datetime.time
def format_time(time):
    # format the time as string
    time = "{0:04d}".format(int(time))
    # extract hh and mm and then convert to time
    hhmm = datetime.time(int(time[0:2]), int(time[2:4]))
    return hhmm

df['SCHEDULED_DEPARTURE'] = \
    df['SCHEDULED_DEPARTURE'].apply(format_time)
df[['SCHEDULED_DEPARTURE']]

SCHEDULED_DEPARTURE
0                  00:05:00
1                  00:10:00
2                  00:20:00
3                  00:20:00
4                  00:25:00
...                     ...
5819074            23:59:00
5819075            23:59:00
5819076            23:59:00
5819077            23:59:00
5819078            23:59:00

[5819079 rows x 1 columns]

## Combining the Date and Time Columns

Now that our dataframe has two columns of datetime datatype — DATETIME and SCHEDULED_DEPARTURE, we can now combine them into a single column. The following code snippet uses the apply() function together with the datetime.combine() function to combine the two specified columns:

In [21]:
from datetime import datetime

df['SCHEDULED_DEPARTURE'] = \
    df.apply(
        lambda r: datetime.combine(r['DATETIME'], r['SCHEDULED_DEPARTURE']), 
        axis=1)
df.sample(5)[['SCHEDULED_DEPARTURE']]

SCHEDULED_DEPARTURE
2144520 2015-05-16 20:25:00
29306   2015-01-02 20:30:00
4692298 2015-10-20 11:44:00
1064176 2015-03-11 09:20:00
5606160 2015-12-18 08:15:00

The SCHEDULED_DEPARTUREcolumn now contains both the date and time of departure:

And now you can easily find all flights departing on specific time:

In [22]:
df[(df['SCHEDULED_DEPARTURE'].dt.month == 12) & 
   (df['SCHEDULED_DEPARTURE'].dt.year == 2015) & 
   (df['SCHEDULED_DEPARTURE'].dt.hour >= 22) & 
   (df['SCHEDULED_DEPARTURE'].dt.minute > 30)]\
   [['FLIGHT_NUMBER','SCHEDULED_DEPARTURE']]

FLIGHT_NUMBER SCHEDULED_DEPARTURE
5355493            214 2015-12-01 22:32:00
5355494           2973 2015-12-01 22:35:00
5355495           5600 2015-12-01 22:35:00
5355496           5627 2015-12-01 22:35:00
5355497            383 2015-12-01 22:35:00
...                ...                 ...
5819074            688 2015-12-31 23:59:00
5819075            745 2015-12-31 23:59:00
5819076           1503 2015-12-31 23:59:00
5819077            333 2015-12-31 23:59:00
5819078            839 2015-12-31 23:59:00

[4863 rows x 2 columns]

## Summary

I hope this article is useful to you when it comes to processing date and time in your Pandas DataFrames. In particular, I have covered:

- How to convert a column to a datetime64 data type after the dataframe is loaded
- How to load a column as a datetime64 object during loading time
- How to find the day of week for a particular date
- How to sort a dataframe based on day of week
- How to combine different columns as a datetime64 object during loading time
- How to combine different columns as a datetime64 object after the dataframe is loaded
- How to convert strings to time format
- How to combine date and time columns
